# Hypothesis Testing

# Statistical Hypothesis Testing

### Building on What We Know:
- Point estimates and confidence intervals are basic tools
- They serve as the **foundation** for another technique: **statistical hypothesis testing**

### What is Statistical Hypothesis Testing?
- A framework for checking if your observed data is different from what you expected
- Helps you decide if patterns in your data are real or just random chance

### Tools Available:
- Python's `scipy.stats` library has many functions for hypothesis tests
- Makes it easy to carry out these tests

# Hypothesis Testing Basics


## The Null Hypothesis:
- Every hypothesis test starts with a **null hypothesis**
- This assumes **nothing interesting is happening** between your variables
- The exact form changes depending on your test:
  - Testing if groups differ? Null says "groups are the same"
  - Example: Testing if your state's average voter age differs from national average
  - Null hypothesis: "There is no difference between the average ages"

## The Alternative Hypothesis:
- This is the opposite of the null hypothesis
- It says **something interesting IS happening**
- Using the same example: "The average age of voters in your state DOES differ from the national average"

## How the Test Works:
1. Start with null and alternative hypotheses
2. Choose a **significance level** (often written as **α**)
3. Run test and get a **p-value**
4. Compare p-value to significance level

## What is Significance Level (α)?
- A probability threshold for rejecting the null hypothesis
- Common choice: α = 0.05 (5%)

## What is P-Value?
- The probability of getting a result as extreme (or more extreme) than what you observed
- This assumes the null hypothesis is true
- If p-value < significance level → reject null hypothesis
- If p-value ≥ significance level → accept null hypothesis

## T-Test:
- A specific type of statistical test
- Used to check if a sample differs significantly from a population
- Or if two samples differ from each other
- Works with numeric data

# One-Sample T-Test

<img src="one_Sample_t-Test.png"></img>

## One-Sample T-Test:
- Checks whether a **sample mean** differs from the **population mean**
- Compares one group to a known standard

## Example:
Let's test whether Minnesota voters have a different average age than all US voters:

- **Population**: All voters in the entire country
- **Sample**: Voters in Minnesota only
- **Question**: Does Minnesota's average voter age differ from the national average?

## What We'll Do:
1. Create dummy age data for the entire US voter population
2. Create a sample of Minnesota voter ages  
3. Run a one-sample t-test to compare them

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt
import math

In [ ]:
np.random.seed(6)

# Generate population from two Poisson distributions
# loc is the minimum value (shift), mu is the mean, size is the number of samples
population_ages1 = stats.poisson.rvs(loc=18, mu=35, size=150000)
population_ages2 = stats.poisson.rvs(loc=18, mu=10, size=100000)

# Combine the two distributions to form the population
population_ages = np.concatenate((population_ages1, population_ages2))

# Generate sample from population using two Poisson distributions
minnesota_ages1 = stats.poisson.rvs(loc=18, mu=30, size=30)
minnesota_ages2 = stats.poisson.rvs(loc=18, mu=10, size=20)

# Combine the two distributions to form the Minnesota sample
minnesota_ages = np.concatenate((minnesota_ages1, minnesota_ages2))

print( population_ages.mean() )
print( minnesota_ages.mean() )

43.000112
39.26


## Setting Up the Test:
- We created Minnesota sample data using a **different distribution** than the population
- This means we **know** the two means are actually different
- Now let's see if the t-test can detect this difference

## Test Parameters:
- **Confidence level**: 95%
- **Null hypothesis**: Minnesota sample comes from the same distribution as the population
- **Alternative hypothesis**: Minnesota sample comes from a different distribution

## Running the Test:
- Use Python's `stats.ttest_1samp()` function
- This function performs a one-sample t-test for us

## What We Expect:
- Since we know the means are different, the test should **reject the null hypothesis**
- This would be the correct result

In [ ]:
# Perform one-sample t-test
stats.ttest_1samp(a = minnesota_ages,               # Sample data
                 popmean = population_ages.mean())  # Pop mean

# sigma = 0.05

TtestResult(statistic=-2.5742714883655027, pvalue=0.013118685425061678, df=49)

In [ ]:
# p-value = 0.013118685425061678
0.05 > 0.013118685425061678

True

## Understanding the Test Results:

### Test Statistic:
- The **t-statistic** = -2.574
- This tells us **how much** the sample mean deviates from the null hypothesis
- Negative value means the sample mean is below what we expected

### Decision Rule:
- If p-value < significance level → reject null hypothesis
- If p-value ≥ significance level → accept null hypothesis
- 0.013118685425061678 < 0.05 → reject null hypothesis

# Two-Sample T-Test

<img src="one_Sample_t-Test.png"></img>

### What It Does:
- Investigates whether the **means of two independent samples** differ from each other
- Compares two groups directly (no population parameter needed)

### Key Differences from One-Sample Test:
- **One-sample**: Compares sample mean to known population parameter
- **Two-sample**: Compares two sample means to each other

### Null Hypothesis:
- The means of both groups are **the same**
- No difference between the two samples

### Alternative Hypothesis:
- The means of the two groups are **different**

### Running the Test:
- Use Python's `stats.ttest_ind()` function
- "ind" stands for "independent" samples

### Example Setup:
- **Sample 1**: Minnesota voter ages (from earlier)
- **Sample 2**: Wisconsin voter ages (we'll generate this)
- **Question**: Do Minnesota and Wisconsin voters have different average ages?

In [ ]:
np.random.seed(12)
# Generate sample from Wisconsin using two Poisson distributions
wisconsin_ages1 = stats.poisson.rvs(loc=18, mu=33, size=30)
wisconsin_ages2 = stats.poisson.rvs(loc=18, mu=13, size=20)

# Combine the two distributions to form the Wisconsin sample
wisconsin_ages = np.concatenate((wisconsin_ages1, wisconsin_ages2))

print( wisconsin_ages.mean() ) # Print Wisconsin sample mean

42.8


In [ ]:
# Perform two-sample t-test
stats.ttest_ind(a= minnesota_ages,
                b= wisconsin_ages,
                equal_var=False)    # Assume samples have equal variance?

TtestResult(statistic=-1.7083870793286842, pvalue=0.09073104343957748, df=97.9724575497005)

In [ ]:
0.05 > 0.09

## Interpreting the Results:

### P-Value:
- **P-value = 0.0907** (about 9%)

### Decision Making:
- **Confidence level**: 95%
- **Significance level**: 5% (α = 0.05)
- **Our p-value**: 9.07%

### The Decision:
- Since **p-value (9.07%) > significance level (5%)** → **Fail to reject** the null hypothesis
- We don't have enough evidence to say Minnesota and Wisconsin voter ages are different

### What This Means:
- The difference we observed could reasonably be due to random chance
- We can't confidently say the two states have different average voter ages
- The evidence isn't strong enough to reject the null hypothesis

# Paired T-Test

<img src="one_Sample_t-Test.png"></img>

### When to Use It:
- For testing differences between **the same group** at different points in time
- Not for independent groups (use regular two-sample t-test for that)

### Real-World Example:
- **Hospital scenario**: Testing if a weight-loss drug works
- **Same patients** measured before and after treatment
- **Question**: Did the patients' weights change significantly?

### Key Difference:
- **Independent t-test**: Compares two different groups
- **Paired t-test**: Compares the same group at two different times

### Running the Test:
- Use Python's `stats.ttest_rel()` function
- "rel" stands for "related" samples

### Example Setup:
- Generate dummy patient weight data
- **Before treatment** weights
- **After treatment** weights
- Test if there's a significant difference

In [ ]:
np.random.seed(11)

# Generate before and after weights
# scale is the standard deviation, loc is the mean, size is the number of samples
# before weights are normally distributed with a mean of 250 pounds
before = stats.norm.rvs(scale=30, loc=250, size=100)

# after weights are normally distributed with a mean decrease of 1.25 pounds
after = before + stats.norm.rvs(scale=5, loc=-1.25, size=100)

# Create a DataFrame to hold the data
weight_df = pd.DataFrame({"weight_before":before,
                          "weight_after":after,
                          "weight_change":after-before})

weight_df.describe()             # Check a summary of the data

,weight_before,weight_after,weight_change
count,100.000000,100.000000,100.000000
mean,250.345546,249.115171,-1.230375
std,28.132539,28.422183,4.783696
min,170.400443,165.913930,-11.495286
25%,230.421042,229.148236,-4.046211
50%,250.830805,251.134089,-1.413463
75%,270.637145,268.927258,1.738673
max,314.700233,316.720357,9.759282


- The summary shows that patients lost about 1.23 pounds on average after treatment.
- Let's conduct a paired t-test to see whether this difference is significant at a 95% confidence level:

In [ ]:
# Perform paired t-test
stats.ttest_rel(a = before,
                b = after)

TtestResult(statistic=2.5720175998568284, pvalue=0.011596444318439857, df=99)

## Test Results:

### Output Breakdown:
- **Test statistic (t)**: 2.572
- **P-value**: 0.0116 (about 1.16%)
- **Degrees of freedom**: 99

### Interpreting the P-Value:
- **P-value = 1.16%**

### Decision Making:
- **Significance level**: Usually 5% (α = 0.05)
- **Our p-value**: 1.16%
- Since **p-value (1.16%) < significance level (5%)** → **Reject the null hypothesis**

### Conclusion:
- We have **strong evidence** that the weight-loss drug works
- The difference in weights before and after treatment is **statistically significant**
- It's very unlikely this difference happened by chance alone

### Degrees of Freedom:
- **df = 99** means we had 100 patients in our study (n - 1 = 99)

# Wrap Up

The t-test is a powerful tool for investigating the differences between sample and population means. T-tests operate on numeric variables; in the next lesson, we'll discuss statistical tests for categorical variables.

### References:
1. https://www.kaggle.com/code/hamelg/python-for-data-24-hypothesis-testing/notebook
2. Seabold, Skipper, and Josef Perktold. “statsmodels: Econometric and statistical modeling with python.” Proceedings of the 9th Python in Science Conference. 2010.